cumulus/pallets/parachain-system/src/lib.rs

```diff
@@ -1218,88 +1256,67 @@ impl<T: Config> Pallet<T> {
 		// channel exists if its MQC head is present in `vfp.hrmp_mqc_heads`.
-		for sender in horizontal_messages.keys() {
+		for sender in horizontal_messages.get_senders() {
 			// A violation of the assertion below indicates that one of the messages submitted
 			// by the collator was sent from a sender that doesn't have a channel opened to
 			// this parachain, according to the relay-parent state.
-			assert!(ingress_channels.binary_search_by_key(sender, |&(s, _)| s).is_ok(),);
+			assert!(ingress_channels.binary_search_by_key(&sender, |&(s, _)| s).is_ok(),);
 		}
 
-		// Second, prepare horizontal messages for a more convenient processing:
-		//
-		// instead of a mapping from a para to a list of inbound HRMP messages, we will have a
-		// list of tuples `(sender, message)` first ordered by `sent_at` (the relay chain block
-		// number in which the message hit the relay-chain) and second ordered by para id
-		// ascending.
-		//
-		// The messages will be dispatched in this order.
-		let mut horizontal_messages = horizontal_messages
-			.into_iter()
-			.flat_map(|(sender, channel_contents)| {
-				channel_contents.into_iter().map(move |message| (sender, message))
-			})
-			.collect::<Vec<_>>();
-		horizontal_messages.sort_by(|a, b| {
-			// first sort by sent-at and then by the para id
-			match a.1.sent_at.cmp(&b.1.sent_at) {
-				cmp::Ordering::Equal => a.0.cmp(&b.0),
-				ord => ord,
-			}
-		});
+		let (messages, hashed_messages) = horizontal_messages.messages();
+		let mut mqc_heads = <LastHrmpMqcHeads<T>>::get();
 
-		let last_mqc_heads = <LastHrmpMqcHeads<T>>::get();
-		let mut running_mqc_heads = BTreeMap::new();
-		let mut hrmp_watermark = None;
+		if messages.is_empty() {
+			Self::check_hrmp_mcq_heads(ingress_channels, &mut mqc_heads);
+			let last_processed_msg =
+				InboundMessageId { sent_at: relay_parent_number, reverse_idx: 0 };
+			LastProcessedHrmpMessage::<T>::put(last_processed_msg);
+			HrmpWatermark::<T>::put(relay_parent_number);
+			return T::DbWeight::get().reads_writes(1, 2);
+		}
 
-		{
-			for (sender, ref horizontal_message) in &horizontal_messages {
-				if hrmp_watermark.map(|w| w < horizontal_message.sent_at).unwrap_or(true) {
-					hrmp_watermark = Some(horizontal_message.sent_at);
-				}
+		let mut last_processed_block = HrmpWatermark::<T>::get();
+		let mut last_processed_msg = InboundMessageId { sent_at: 0, reverse_idx: 0 };
+		for (sender, msg) in messages {
+			if msg.sent_at > last_processed_msg.sent_at && last_processed_msg.sent_at > 0 {
+				last_processed_block = last_processed_msg.sent_at;
+			}
+			last_processed_msg.sent_at = msg.sent_at;
 
-				running_mqc_heads
-					.entry(sender)
-					.or_insert_with(|| last_mqc_heads.get(sender).cloned().unwrap_or_default())
-					.extend_hrmp(horizontal_message);
+			mqc_heads.entry(*sender).or_default().extend_hrmp(msg);
+		}
+		<LastHrmpMqcHeads<T>>::put(mqc_heads.clone());
```

Comment:
```suggestion
		<LastHrmpMqcHeads<T>>::put(&mqc_heads);
```
Refs implement EncodeLike.